File: src/product/Alloy.java
package product;

import utils.Codes;

public class Alloy implements Codes {
    protected double price;
    protected double platinum;
    protected double iron;
    protected double copper;
    protected double electricity;

    public Alloy(double price, double platinum, double iron, double copper, double electricity) {
        this.price = price;
        this.platinum = platinum;
        this.iron = iron;
        this.copper = copper;
        this.electricity = electricity;
    }

    /**
     * Calculates the total resource quantities to make a number of alloys
     * @param alloy amounts
     * @return an array with the quantity of each resource required
     */
    public double[] calc(int alloy) {
        double[] quant = new double[4];

        quant[PLATINUM] = platinum     * alloy;
        quant[IRON] = iron         * alloy;
        quant[COPPER] = copper       * alloy;
        quant[ELECTRICITY] = electricity  * alloy;

        return quant;
    }
}

File: src/product/Alloy.kt
package product

class Alloy (private val price: Double,
             private val platinum: Double,
             private val iron: Double,
             private val copper: Double,
             private val electricity: Double) {

    private val PLATINUM = "platinum"
    private val IRON = "iron"
    private val COPPER = "copper"
    private val ELECTRICITY = "electricity"


    /**
     * Calculates the total resource quantities to make a number of alloys
     * @param alloy amounts
     * @return an array with the quantity of each resource required
     */
    fun calc(alloy: Double): Map<String, Double> {
        return mapOf(
                PLATINUM    to Math.abs(platinum * alloy),
                IRON        to Math.abs(iron * alloy),
                COPPER      to Math.abs(copper * alloy),
                ELECTRICITY to Math.abs(electricity * alloy))
    }

    fun price(quant: Double) : Double{
        if (quant <= 0.0) return 0.0
        return quant * price
    }
}